Sources/AsyncQueue/ActorQueue.swift

```diff
@@ -0,0 +1,116 @@
+// MIT License
+//
+// Copyright (c) 2022 Dan Federman
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/// A queue that executes asynchronous tasks enqueued from a nonisolated context.
+/// Tasks are guaranteed to begin executing in the order in which they are enqueued. However, if a task suspends it will allow tasks that were enqueued to begin executing.
+/// Asynchronous tasks sent to this queue execute as they would in an `actor` type, allowing for re-entrancy and non-FIFO behavior when an individual task suspends.
+/// - Warning: Execution order is not guaranteed unless the enqueued tasks interact with a single `actor` instance.
```

Comment:
It may make sense to attach the actor itself to the queue? Since other ways of doing this seem to hit against the limitations of the Swift type system...

I don't foresee a sensible use case for this where the same queue is used for several different actors.

Reply:
I thought about trying to associate a single actor with a queue, but fell short coming up with a good way to do it. As-written, I expect folk will utilize an `ActorQueue` as follows:

```
actor MyActor {

    nonisolated
    func send(input: Input) {
        queue.async {
            await self.process(input: input)
        }
    }

    private func process(input: Input) {
        // Process the ordered input on the actor queue
    }

    private let queue = ActorQueue()

}
```

This isn't a whole lot of boilerplate, and I haven't come up with something cleaner.

I'm not terribly worried about folk mis-using this API assuming we get the documentation right. Folk coming to this repo are likely trying to solve a very specific problem and will use this code as intended (again, assuming the docs are good).

That said, I'm open to ideas. I've been unable to come up with anything better thus far than what we've got here, but that doesn't mean there aren't better options!

Reply:
For what it's worth, I tried a ton of different shenanigans to get rid of this warning but came up short. What I need is a way to signal the semaphore after we've suspended the current context's execution to `await` the task. Basically I need something like:

```
// Signal the semaphore once we've begun waiting for the `task`.
deferUntilSuspension {
    await semaphore.signal()
}
await task()
```

But that isn't an existing language feature!

Reply:
A possible alternative would be to store the actor in the queue class and then also pass it to the execution blocks as a parameter. Decouples things a bit more than the raw blocks here.

Reply:
I'm not convinced that would work, since any suspension to signal the semaphore would potentially result in the actor methods executing in an unexpected order 🤔. See https://github.com/dfed/swift-async-queue/pull/3/commits/0565678b069a5c6d195113e261ecf449d626cbfa#r1033680644

Reply:
Resolving this thread since it's been idle for a bit and neither of us have come up with good alternatives. Feel free to un-resolve.